### PR TITLE
[SPARK-49836][SQL][SS] Fix possibly broken query when window is provided to window/session_window fn

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR fixes the correctness issue about losing operators during analysis - it happens when window is provided to window()/session_window() function.

The rule `TimeWindowing` and `SessionWindowing` are responsible to resolve the time window functions. When the window function has `window` as parameter (time column) (in other words, building time window from time window), the rule wraps window with WindowTime function so that the rule ResolveWindowTime will further resolve this. (And TimeWindowing/SessionWindowing will resolve this again against the result of ResolveWindowTime.)

The issue is that the rule uses "return" for the above, which intends to have "early return" as the other branch is too long compared to this branch. This unfortunately does not work as intended - the intention is just to go out of current local scope (mostly end of curly brace), but it seems to break the loop of execution in "outer" side.
(I haven't debugged further but it's simply clear that it doesn't work as intended.)

Quoting from Scala doc:

> Nonlocal returns are implemented by throwing and catching scala.runtime.NonLocalReturnException-s.

It's not super clear where NonLocalReturnException is caught in the call stack; it might exit the execution for much broader scope (context) than expected. And it's finally deprecated in Scala 3.2 and likely be removed in future.

https://dotty.epfl.ch/docs/reference/dropped-features/nonlocal-returns.html

Interestingly it does not break every query for chained time window aggregations. Spark already has several tests with DataFrame API and they haven't failed. The reproducer in community report is using SQL statement - where each aggregation is considered as subquery.

This PR fixes the rule to NOT use early return and instead have a huge if else.

### Why are the changes needed?

Described in above.

### Does this PR introduce _any_ user-facing change?

Yes, this fixes the possible query breakage. The impacted workloads may not be very huge as chained time window aggregations is an advanced usage, and it does not break every query for the usage.

### How was this patch tested?

New UTs.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
https://dotty.epfl.ch/docs/reference/dropped-features/nonlocal-returns.html

Going further, this is deprecated in newer Scala version (though it's like 3.2 which Spark wouldn't catch up for years) and will be removed. This clearly indicates that programmers have problems with this and they want to make this be much more explicit rather than just the same with local return.

**Comment by HeartSaVioR** [comment]
cc. @cloud-fan Would you mind take a look at this? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTimeWindows.scala [hunk 1]
```diff
@@ -87,85 +87,86 @@ object TimeWindowing extends Rule[LogicalPlan] {
 
         val window = windowExpressions.head
 
+        // time window is provided as time column of window function, replace it with WindowTime
```
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTimeWindows.scala:90]: For reviewer: except this code comment, the only change is to remove return and replace it with a huge if-else statement.

**Comment by HeartSaVioR** [comment]
cc. @viirya

**Review by viirya — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master/3.5/3.4.

**Pull request closed without merging**